lib/validators/validation.rb

```diff
@@ -0,0 +1,214 @@
+      end
+
+      if version == :dstu2
+        guess_dstu2_profile(resource)
+      elsif version == :stu3
+        guess_stu3_profile(resource)
+      elsif version == :r4
+        guess_r4_profile(resource)
+      end
+    end
+
+    def self.guess_dstu2_profile(resource)
+      return if resource.blank?
+
+      candidates = RESOURCES[:dstu2][resource.resourceType]
+      return if candidates.blank?
+
+      # Special cases where there are multiple profiles per Resource type
+      if resource.resourceType == 'Observation'
+        if resource&.code&.coding&.any? { |coding| coding&.code == '72166-2' }
+          return DEFINITIONS[ARGONAUT_URIS[:smoking_status]]
+        elsif resource&.category&.coding&.any? { |coding| coding&.code == 'laboratory' }
+          return DEFINITIONS[ARGONAUT_URIS[:observation_results]]
+        elsif resource&.category&.coding&.any? { |coding| coding&.code == 'vital-signs' }
+          return DEFINITIONS[ARGONAUT_URIS[:vital_signs]]
+        end
+      elsif resource.resourceType == 'CarePlan'
+        if resource&.category&.any? { |category| category&.coding&.any? { |coding| coding&.code == 'careteam' } }
+          return DEFINITIONS[ARGONAUT_URIS[:care_team]]
+        else
+          return DEFINITIONS[ARGONAUT_URIS[:care_plan]]
+        end
+      end
+
+      # Otherwise, guess the first profile that matches on resource type
+      candidates.first
+    end
+
+    def self.guess_stu3_profile(resource)
+      return if resource.blank?
+
+      candidates = RESOURCES[:stu3][resource.resourceType]
+      return if candidates.blank?
+
+      # Special cases where there are multiple profiles per Resource type
+      if resource.resourceType == 'ExplanationOfBenefit'
+        if resource&.type&.coding&.any? { |coding| coding.code == 'CARRIER' }
+          return DEFINITIONS[BLUEBUTTON_URIS[:carrier]]
+        elsif resource&.type&.coding&.any? { |coding| coding.code == 'DME' }
+          return DEFINITIONS[BLUEBUTTON_URIS[:dme]]
+        elsif resource&.type&.coding&.any? { |coding| coding.code == 'HHA' }
+          return DEFINITIONS[BLUEBUTTON_URIS[:hha]]
+        elsif resource&.type&.coding&.any? { |coding| coding.code == 'HOSPICE' }
+          return DEFINITIONS[BLUEBUTTON_URIS[:hospice]]
+        elsif resource&.type&.coding&.any? { |coding| coding.code == 'INPATIENT' }
+          return DEFINITIONS[BLUEBUTTON_URIS[:inpatient]]
+        elsif resource&.type&.coding&.any? { |coding| coding.code == 'OUTPATIENT' }
+          return DEFINITIONS[BLUEBUTTON_URIS[:outpatient]]
+        elsif resource&.type&.coding&.any? { |coding| coding.code == 'PDE' }
+          return DEFINITIONS[BLUEBUTTON_URIS[:pde]]
+        elsif resource&.type&.coding&.any? { |coding| coding.code == 'SNF' }
+          return DEFINITIONS[BLUEBUTTON_URIS[:snf]]
+        end
+      end
+
+      # Otherwise, guess the first profile that matches on resource type
+      candidates.first
+    end
+
+    def self.guess_r4_profile(resource)
```

Comment:
I assume this profile matching stuff is copied in from somewhere? If so, can we put a comment sourcing where it's from? (might help with future maintainability)